optuna/_gp/gp.py

```diff
@@ -177,25 +180,33 @@ def loss_func(raw_params: np.ndarray) -> tuple[float, np.ndarray]:
             kernel_scale=torch.exp(raw_params_tensor[n_params]),
             noise_var=torch.exp(raw_params_tensor[n_params + 1]) + minimum_noise,
         )
+        if deterministic_objective:
+            params.noise_var = torch.tensor(minimum_noise, dtype=torch.float64)
         loss = -marginal_log_likelihood(
             torch.from_numpy(X), torch.from_numpy(Y), torch.from_numpy(is_categorical), params
         ) - log_prior(params)
         loss.backward()  # type: ignore
+        # scipy.minimize requires all the gradients to be zero for termination.
+        assert not deterministic_objective or raw_params_tensor.grad[n_params + 1] == 0
         return loss.item(), raw_params_tensor.grad.detach().numpy()  # type: ignore
 
     # jac=True means loss_func returns the gradient for gradient descent.
-    res = so.minimize(loss_func, initial_raw_params, jac=True)
+    res = so.minimize(
+        loss_func, initial_raw_params, jac=True, method="l-bfgs-b", options={"gtol": gtol}
```

Comment:
```suggestion
        # We need a high gtol value because the loss_func can have high numerically errors.
        loss_func, initial_raw_params, jac=True, method="l-bfgs-b", options={"gtol": gtol}
```